### PR TITLE
cleanup entr_detr funcs

### PR DESCRIPTION
**Pull request by haakon-e**

### CODE REVIEW

**Comment by haakon-e** [comment]
do you know why juliaformatter is failing, @charleskawczynski?

File: src/closures/entr_detr.jl [hunk 1]
```diff
@@ -14,46 +14,63 @@ function compute_turbulent_entrainment(param_set, εδ_model_vars)
...
-    # clip vertical velocity
-    if Δw < 0.0
-        Δw -= w_min
-    else
-        Δw += w_min
-    end
-
+function get_MdMdz(εδ_model_vars)
     MdMdz = max(εδ_model_vars.dMdz / max(εδ_model_vars.M, 1e-12), 0.0)
     MdMdz = max(-εδ_model_vars.dMdz / max(εδ_model_vars.M, 1e-12), 0.0)
```
Comment (haakon-e) [inline src/closures/entr_detr.jl:44]: I tried to trace down why this function call is made twice. It seems like it was inadvertently introduced in this commit: https://github.com/CliMA/TurbulenceConvection.jl/commit/d37ebb224c6d22dbebc95a32d78dd6c08f5a488e#diff-5697e07122904684a344ca1589e54a3e13e7faee3517cfd69f4b8116349a5534R68-R69
thoughts, @yairchn?

Comment (haakon-e) [reply src/closures/entr_detr.jl:44]: In the old code, the first line was `entr_MdMdz` and the second `detr_MdMdz`. See these lines: https://github.com/CliMA/TurbulenceConvection.jl/commit/d37ebb224c6d22dbebc95a32d78dd6c08f5a488e#diff-10852a8223babe2ea765dcd9b536c68ccd42c65b5053089a7e8c78f85527e8baL128-L129

File: src/closures/entr_detr.jl [hunk 2]
```diff
@@ -14,46 +14,63 @@ function compute_turbulent_entrainment(param_set, εδ_model_vars)
...
     l = SA.SVector(l_1, l_2)
     return lamb_smooth_minimum(l, 0.1, 0.0005)
-
 end
 
-function preprocess_inputs(param_set, εδ_model_vars, Δw)
+function get_Δw(param_set, εδ_model_vars)
+    Δw = εδ_model_vars.w_up - εδ_model_vars.w_en
+    Δw += copysign(CPEDMF.w_min(param_set), Δw)
+    return Δw
```
Comment (haakon-e) [inline src/closures/entr_detr.jl:33]:
```suggestion
    return Δw
end
```

File: src/closures/entr_detr.jl [hunk 3]
```diff
@@ -63,39 +80,33 @@ Cohen et al. (JAMES, 2020), given:
  - `εδ_model_type`  :: MDEntr - Moisture deficit entrainment closure
 """
 function entr_detr(param_set, εδ_model_vars, εδ_model_type::MDEntr)
-
-    γ_lim = ICP.area_limiter_scale(param_set)
-    β_lim = ICP.area_limiter_power(param_set)
+    dim_scale = dimensional_part(param_set, εδ_model_vars)
+    area_limiter = max_area_limiter(param_set, εδ_model_vars)
+
```
Comment (haakon-e) [inline src/closures/entr_detr.jl:85]:
```suggestion

```

File: src/closures/entr_detr.jl [hunk 4]
```diff
@@ -63,39 +80,33 @@ Cohen et al. (JAMES, 2020), given:
...
+    
+    # Moisture deficit closure
+    Δw = get_Δw(param_set, εδ_model_vars)
+    D_ε, D_δ, M_δ, M_ε = nondimensional_exchange_functions(param_set, Δw, εδ_model_vars)
     c_ε = CPEDMF.c_ε(param_set)
-    c_t = CPEDMF.c_t(param_set)
+    c_δ = get_c_δ(param_set, εδ_model_vars)
+    nondim_ε = (c_ε * D_ε + c_δ * M_ε)
+    nondim_δ = (c_ε * D_δ + c_δ * M_δ)
+
```
Comment (haakon-e) [inline src/closures/entr_detr.jl:93]:
```suggestion

```

File: src/closures/entr_detr.jl [hunk 5]
```diff
@@ -63,39 +80,33 @@ Cohen et al. (JAMES, 2020), given:
...
-    Δw, MdMdz, c_δ = preprocess_inputs(param_set, εδ_model_vars, Δw)
-
-    D_ε, D_δ, M_δ, M_ε = nondimensional_exchange_functions(param_set, Δw, Δb, εδ_model_vars)
-
-    λ = compute_inverse_timescale(Δb, Δw, param_set, εδ_model_vars)
-
+    # dynamic entrainment / detrainment
+    ε_dyn = dim_scale * nondim_ε + drybubble_adj
+    δ_dyn = dim_scale * nondim_δ + drybubble_adj + area_limiter
+
```
Comment (haakon-e) [inline src/closures/entr_detr.jl:101]:
```suggestion

```

File: src/closures/entr_detr.jl [hunk 6]
```diff
@@ -105,38 +116,24 @@ network to predict the non-dimensional component of dynamical entrainment:
...
-    # turbulent entrainment
-    ε_turb, K_ε = compute_turbulent_entrainment(param_set, εδ_model_vars)
-
-    logistic_term = (2.0 - 1.0 / (1 + exp(-γ_lim * (εδ_model_vars.max_area - εδ_model_vars.a_up))))
-    max_area_limiter = λ / Δw * (logistic_term^β_lim - 1.0)
-
-    nn_arc = (4, 2, 2) # number of inputs, number of neurons, number of outputs
+    dim_scale = dimensional_part(param_set, εδ_model_vars)
+    area_limiter = max_area_limiter(param_set, εδ_model_vars)
+
```
Comment (haakon-e) [inline src/closures/entr_detr.jl:121]:
```suggestion

```

File: src/closures/entr_detr.jl [hunk 7]
```diff
@@ -105,38 +116,24 @@ network to predict the non-dimensional component of dynamical entrainment:
...
 
-    pi_1 = εδ_model_vars.updraft_top * Δb / Δw^2
-    pi_2 = εδ_model_vars.tke / Δw^2
-    non_dim_functions = nn_model([pi_1, pi_2, εδ_model_vars.a_up, εδ_model_vars.RH_up - εδ_model_vars.RH_en])
-    ε_dyn = (λ / Δw) * non_dim_functions[1]
-    δ_dyn = (λ / Δw) * non_dim_functions[2] + max_area_limiter
+    # dynamic entrainment / detrainment
+    ε_dyn = dim_scale * nondim_ε
+    δ_dyn = dim_scale * nondim_δ + area_limiter
+
```
Comment (haakon-e) [inline src/closures/entr_detr.jl:134]:
```suggestion

```

**Review by haakon-e — COMMENTED** [review]
formatting

File: src/closures/entr_detr.jl [hunk 8]
```diff
@@ -14,46 +14,63 @@ function compute_turbulent_entrainment(param_set, εδ_model_vars)
...
     l_2 = abs(Δb / Δw)
     l = SA.SVector(l_1, l_2)
     return lamb_smooth_minimum(l, 0.1, 0.0005)
-
 end
 
-function preprocess_inputs(param_set, εδ_model_vars, Δw)
+function get_Δw(param_set, εδ_model_vars)
+    Δw = εδ_model_vars.w_up - εδ_model_vars.w_en
+    Δw += copysign(CPEDMF.w_min(param_set), Δw)
```
Comment (charleskawczynski) [inline src/closures/entr_detr.jl:30]: Let's (statically) get `FT` and wrap this to ensure it's type-stable.

**Review by charleskawczynski — APPROVED** [review]
Overall the code changes look fine to me. I left one comment on ensuring type stability.

File: src/closures/entr_detr.jl [hunk 9]
```diff
@@ -14,46 +14,63 @@ function compute_turbulent_entrainment(param_set, εδ_model_vars)
...
+    return area_limiter
+end
+
+function non_dimensional_groups(param_set, εδ_model_vars)
+    Δw = get_Δw(param_set, εδ_model_vars)
+    Δb = εδ_model_vars.b_up - εδ_model_vars.b_en
+    pi_1 = εδ_model_vars.updraft_top * Δb / Δw^2
+    pi_2 = εδ_model_vars.tke / Δw^2
+    pi_3 = √(εδ_model_vars.a_up)
+    pi_4 = εδ_model_vars.RH_up - εδ_model_vars.RH_en
```
Comment (yairchn) [inline src/closures/entr_detr.jl:68]: could we use the letter '\pi' here?

File: src/closures/entr_detr.jl (hunk 1 above)
Comment (yairchn) [reply src/closures/entr_detr.jl:44]: looks to me that there was a mistake in the past PR that unified 'entr_MdMdz' and  'detr_MdMdz'.
What we want is to compute the product 'MdMdz', take the positive part of it to entrainment (MdMdz_entr) and the negative (in absolute value) for detrainment (MdMdz_detr).

**Review by costachris — APPROVED** [review]

File: src/closures/entr_detr.jl [hunk 10]
```diff
@@ -14,46 +14,63 @@ function compute_turbulent_entrainment(param_set, εδ_model_vars)
...
+end
+
+function non_dimensional_groups(param_set, εδ_model_vars)
+    Δw = get_Δw(param_set, εδ_model_vars)
+    Δb = εδ_model_vars.b_up - εδ_model_vars.b_en
+    Π_1 = Δw^2 / (εδ_model_vars.zc_i * Δb)
+    Π_2 = Δw^2 / εδ_model_vars.tke
+    Π_3 = √(εδ_model_vars.a_up)
+    Π_4 = εδ_model_vars.RH_up - εδ_model_vars.RH_en
+    return SA.SVector(Π_1, Π_2, Π_3, Π_4)
```
Comment (costachris) [inline src/closures/entr_detr.jl:69]: Is there a reason to use `SA.SVector` instead of just a vector? ie. `return [Π_1, Π_2, Π_3, Π_4]`

**Comment by haakon-e** [comment]
Could you have a quick look at https://github.com/CliMA/TurbulenceConvection.jl/pull/648/commits/ef518c1f9a8f4388fafb8f2d089626d2e5f3a1c3, @charleskawczynski ?

File: src/closures/entr_detr.jl [hunk 11]
```diff
@@ -2,9 +2,10 @@
 using Flux
 
 function compute_turbulent_entrainment(param_set, εδ_model_vars)
-
-    c_t = CPEDMF.c_t(param_set)
+    FT = eltype(εδ_model_vars)
+    c_t = FT(CPEDMF.c_t(param_set))
     K_ε = εδ_model_vars.a_up * c_t * sqrt(max(εδ_model_vars.tke, 0.0)) * εδ_model_vars.R_up
+
```
Comment (charleskawczynski) [inline src/closures/entr_detr.jl:8]: There's a `2.0` and `^2.0` below that you could change, too

File: src/closures/entr_detr.jl [hunk 12]
```diff
@@ -14,46 +15,67 @@ function compute_turbulent_entrainment(param_set, εδ_model_vars)
     return (ε_turb, K_ε)
 end
 
-function compute_inverse_timescale(Δb::FT, Δw::FT, param_set::APS, εδ_model_vars) where {FT}
-
+function compute_inverse_timescale(param_set, εδ_model_vars)
+    Δb = εδ_model_vars.b_up - εδ_model_vars.b_en
+    Δw = get_Δw(param_set, εδ_model_vars)
```
Comment (charleskawczynski) [inline src/closures/entr_detr.jl:20]: We could type cast some constants in here, too.

**Review by charleskawczynski — APPROVED** [review]
Overall LGTM

**Pull request merged**